File: BionicLimbSimulation_unity/Assets/Scripts/Core/EMGDataBuffer.cs
using System;

namespace BionicLimb.Core
{
    /// <summary>
    /// Plain C# data container for a loaded .myo.csv recording.
    /// Spec §6.2
    /// </summary>
    public class EMGDataBuffer
    {
        public double[] Timestamps { get; }       // millisecond offsets
        public float[,] Channels { get; }          // [sampleCount, 8]
        public string[] GestureLabels { get; }     // per-sample gesture name
        public int[] GestureIds { get; }            // per-sample gesture id
        public int SampleRate { get; }
        public int SampleCount { get; }

        public EMGDataBuffer(
            double[] timestamps,
            float[,] channels,
            string[] gestureLabels,
            int[] gestureIds,
            int sampleRate)
        {
            Timestamps = timestamps;
            Channels = channels;
            GestureLabels = gestureLabels;
            GestureIds = gestureIds;
            SampleRate = sampleRate;
            SampleCount = timestamps.Length;
        }

        /// <summary>Returns all fields for a single sample index.</summary>
        public EMGSample GetSampleAt(int index)
        {
            if (index < 0 || index >= SampleCount)
                throw new ArgumentOutOfRangeException(nameof(index));

            float[] ch = new float[8];
            for (int i = 0; i < 8; i++) ch[i] = Channels[index, i];

            return new EMGSample
            {
                Index = index,
                TimestampMs = Timestamps[index],
                Channels = ch,
                GestureLabel = GestureLabels[index],
                GestureId = GestureIds[index]
            };
        }

        /// <summary>
        /// Returns a [windowSize, 8] slice centred on centerIndex.
        /// Pads with zeros if window extends past buffer edges.
        /// Used by classifiers (§6.2).
        /// </summary>
        public float[,] GetWindowAt(int centerIndex, int windowSize)
        {
            var window = new float[windowSize, 8];
            int half = windowSize / 2;
            for (int w = 0; w < windowSize; w++)
            {
                int src = centerIndex - half + w;
                if (src < 0 || src >= SampleCount) continue;
                for (int c = 0; c < 8; c++)
                    window[w, c] = Channels[src, c];
            }
            return window;
        }
    }

    public struct EMGSample
    {
        public int Index;
        public double TimestampMs;
        public float[] Channels;
        public string GestureLabel;
        public int GestureId;
    }
}

File: bionic-limb-threejs/src/core/EMGDataBuffer.ts
import type { MyoSample } from './MyoFileParser.js';

export class EMGDataBuffer {
  readonly sampleCount: number;
  readonly channelCount: number = 8;
  readonly sampleRateHz: number;
  readonly samples: readonly MyoSample[];

  constructor(samples: MyoSample[], sampleRateHz: number) {
    this.samples = samples;
    this.sampleCount = samples.length;
    this.sampleRateHz = sampleRateHz;
  }

  getSampleAt(index: number): Float32Array {
    if (index < 0 || index >= this.sampleCount) {
      return new Float32Array(8);
    }
    return this.samples[index].channels;
  }

  getWindowAt(index: number, windowSize: number): Float32Array[] {
    const window: Float32Array[] = [];
    for (let i = 0; i < windowSize; i++) {
      const sampleIdx = index - windowSize + 1 + i;
      if (sampleIdx < 0) {
        window.push(new Float32Array(8));
      } else {
        window.push(this.getSampleAt(sampleIdx));
      }
    }
    return window;
  }
}
